rids/rids.py

```diff
@@ -14,114 +14,139 @@
+from rids.rules.ruleset import RuleSet
 
 FLAGS = flags.FLAGS
 flags.DEFINE_string('host_ip', None,
-                    'The IP of the host process, for ignoring direct requests')
-flags.DEFINE_string('config_path', 'config.json',
-                    'Path where IOC configuration can be found; uses config.json by default')
-
-
-def retrieve_bad_ips():
-  bad_ips = {}
-  with open(FLAGS.config_path, 'r') as f:
-    config = json.load(f)
-    for source in config['ioc_sources']:
-      # Download a fresh version of each IOC source in the config.
-      with urllib.request.urlopen(source['url']) as ioc_data:
-        # TODO check format of ioc source, branch to different parsers
-        # For now, the only source format is newline-seprated bad IPv4 addresses
-        # TODO perform parsing in a separate function from config handling
-        for line in ioc_data.readlines():
-          line = line.strip().decode('utf-8')
-          if not line: continue
-          bad_ip = str(ipaddress.ip_address(line))
-          if bad_ip not in bad_ips:
-            bad_ips[bad_ip] = [source['name']]
-          else:
-            bad_ips[bad_ip].append(source['name'])
-  return bad_ips
-
-
-def get_external_ip():
-  with urllib.request.urlopen('https://ipinfo.io/ip') as my_ip:
-    return my_ip.read().decode("utf-8").strip()
+                    'The IP of the host process, helps determine whether an IP '
+                    'is a client or a remote endpoint.')
+flags.DEFINE_string('config_path', '/etc/rids/config.json',
+                    'file path indicating where IOC configuration can be found')
+flags.DEFINE_string('eventlog_path', None,
+                    'file path indicating where to append new event logs')
 
 
 def main(argv):
+  """Main entry point of the app.  Also bound to CLI `rids` by setuptools."""
   if len(argv) > 1:
     raise app.UsageError('Too many command-line arguments.')
 
+  eventlog_path = _get_eventlog_path()
+  logging.basicConfig(level=logging.DEBUG,
+                      format='%(asctime)s %(levelname)-8s %(message)s',
+                      datefmt='%m-%d %H:%M',
+                      filename=eventlog_path,
+                      filemode='a')
+
+  config = _load_config()
+  ruleset : RuleSet = iocs.fetch_iocs(config)
+
+  host_ip = _get_host_ip()
+  print(f'Using {host_ip} as host IP address')
+
+  event_queue = queue.Queue()
+  loop = asyncio.get_event_loop()
+  with ThreadPoolExecutor() as executor:
+    loop.run_in_executor(
+        executor,
+        partial(_inspect_tls_traffic, host_ip, ruleset, event_queue))
+    loop.run_in_executor(
+        executor,
+        partial(_inspect_remote_endpoints, host_ip, ruleset, event_queue))
+  
+    while True:
+      # We just log the suspicious events for now, but here is where we could
+      # do post-processing and/or share sightings of known / unknown threats.
+      event = event_queue.get()
+      logging.log(json.dumps(event))
```

Comment:
We are conflating a match and a log entry. At some point we should separate those things, but we can do that later.

Reply:
Yes, they aren't necessarily the same.